### PR TITLE
fix(determinism): Separate the MoE aux-loss fusion flag so --deterministic-mode guards only the op that breaks it

### PR DESCRIPTION
**Pull request by ZhiyuLi-Nvidia**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Separates the MoE aux-loss fusion from the TopK routing fusion, so `--deterministic-mode` guards only the op that actually breaks determinism.

Determinism breaks with Transformer Engine's fused MoE aux-loss kernel, so `--deterministic-mode` has to reject it. `moe_router_fusion` is too coarse to express that — one flag gates two things:

- `switch_load_balancing_loss_func` and `compute_routing_scores_for_aux_loss` — the **aux loss**, which reaches the backward through `MoEAuxLossAutoScaler.apply`, so a nonzero `--moe-aux-loss-coeff` puts any wobble in it straight into the gradients
- `topk_routing_with_score_function` — **TopK routing**, which only selects experts

Only the first is implicated. Requiring the whole flag off would also disable the fused routing path, costing throughput for something that is not the problem.

So: add `moe_router_aux_loss_fusion`, point the four aux-loss call sites at it, leave the three routing sites on `moe_router_fusion`, and have `--deterministic-mode` require only the new flag off.

**Compatibility.** The new flag defaults to `None`, meaning "follow `moe_router_fusion`", so existing configs are unaffected — `moe_router_fusion=True` still enables both fusions and still trips the assert. Keeping fused routing under `--deterministic-mode` means setting `moe_router_aux_loss_fusion=False` explicitly.

Resolution happens on read, via `moe_router_aux_loss_fusion_enabled`, not in `__post_init__`: callers flip `moe_router_fusion` on an already-constructed config (`tests/unit_tests/transformer/moe/test_aux_loss.py` does this to exercise its fused leg), and resolving at construction would freeze the inherited value. `apply_determinism_to_args` applies the same rule to `args`, which has no config yet.

**Evidence.** Measured locally. Running the router on identical input in one process, 8 repeats per aux-loss type, comparing the aux loss bitwise (1 GPU, `moe_aux_loss_coeff=1.0`, `moe_router_dtype=fp32`):

| aux-loss type | unfused | fused |
|---|---|---|
| `aux_loss` | 0/7 mismatches | **7/7 mismatches** |
| `seq_aux_loss` | 0/7 mismatches | **5/7 mismatches** |
| `global_aux_loss` | 0/7 mismatches | **5/7 mismatches** |

The unfused path is bit-identical on every repeat; the fused path is not, for all three types.

Two limits. The mismatch was in the aux-loss **value** — the router-weight gradient stayed bit-identical in every fused repeat — so this does not by itself show the wobble reaching the weights; the path for that is `MoEAuxLossAutoScaler.apply` putting the aux loss in the backward, amplified by `--moe-router-enable-expert-bias`. And a 2-node, 5-step training A/A at the production `moe_aux_loss_coeff=1e-4` came back identical, which is consistent with a coefficient four orders of magnitude smaller.

The fused **routing** path is unmeasured in either direction. This change stops requiring it off on the grounds that nothing has been reported against it.

## Issue tracking

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6967/commits/0036ffa7399150fbf117b02ccc0ae383d05bf6cf

**Review by yaox12 — APPROVED** [review]

**Review by Connor-XY — APPROVED** [review]

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6967/commits/1b845d629e325f0101a97a7a980ad03f34392418

**Review by Phlip79 — APPROVED** [review]

File: megatron/core/transformer/moe/router.py [hunk 1]
```diff
@@ -811,7 +811,7 @@ def routing(self, logits: torch.Tensor, padding_mask: Optional[torch.Tensor] = N
                 logits,
                 self.topk,
                 self.score_function,
-                fused=self.config.moe_router_fusion,
```
Comment (yashaswikarnati) [inline megatron/core/transformer/moe/router.py:816]: would we still need moe_router_fusion?

File: docs/user-guide/deterministic-training.md [hunk 2]
```diff
@@ -43,6 +43,7 @@ Checked against the parsed `args` Namespace in `apply_determinism_to_args`. Inco
 |---|---|
 | `--cross-entropy-loss-fusion` | Must be off — asserted (fused CE is non-deterministic); drop the flag yourself |
 | `--tp-comm-overlap` | Must be off — asserted (the overlap path is not bit-exact); drop the flag yourself |
+| `moe_router_aux_loss_fusion` | Must be off — asserted (the fused aux loss reaches the backward); follows `moe_router_fusion` when unset |
```
Comment (jaredcasper) [inline docs/user-guide/deterministic-training.md]: What does "the fused aux loss reaches the backward" mean?

File: megatron/core/transformer/transformer_config.py [hunk 3]
```diff
@@ -1514,6 +1525,17 @@ def _validate_cp_layouts(self) -> None:
                 "linear_cp_layout='contiguous' with context parallelism does not yet support MTP."
             )
 
+    @property
+    def moe_router_aux_loss_fusion_enabled(self) -> bool:
```
Comment (jaredcasper) [inline megatron/core/transformer/transformer_config.py]: Why not just set this to moe_router_fusion in `__post_init__` when it isn't set?

File: megatron/core/transformer/moe/router.py (hunk 1 above)
Comment (ZhiyuLi-Nvidia) [reply megatron/core/transformer/moe/router.py:816]: > would we still need moe_router_fusion?

Were you asking if we need to keep moe_router_fusion flag?
Then the answer is yes.

> moe_router_fusion is too coarse to express that — one flag gates two things:
> 
> * switch_load_balancing_loss_func and compute_routing_scores_for_aux_loss — the aux loss, which reaches the backward through MoEAuxLossAutoScaler.apply, so a nonzero --moe-aux-loss-coeff puts any wobble in it straight into the gradients
> * topk_routing_with_score_function — TopK routing, which only selects experts

The idea is to separate the MoE aux-loss fusion using an independent key, as this is the only fusion op break determinism.

This is also helpful to keep back compatibility.

**Review by maanug-nv — APPROVED** [review]

File: docs/user-guide/deterministic-training.md (hunk 2 above)
Comment (ZhiyuLi-Nvidia) [reply docs/user-guide/deterministic-training.md]: Sorry, that was a confusing way to put it.

What I meant: the aux loss isn't just a number we log. It takes part in the backward pass and ends up in the router gradients. 

But that framing was too complex. The simpler point is that TE's fused aux-loss kernel is not deterministic, full stop. 

I've reworded the doc row and the code comment to just say the kernel is non-deterministic.

Verification:
Same input, 8 runs: a different aux loss on 7 of 7 repeats for aux_loss, and 5 of 7 for seq_aux_loss and global_aux_loss. The unfused version gave the same answer every time. That alone is enough for `--deterministic-mode` to require it off.

File: megatron/core/transformer/transformer_config.py (hunk 3 above)
Comment (ZhiyuLi-Nvidia) [reply megatron/core/transformer/transformer_config.py]: Thank you @jaredcasper, Followed your advice — `__post_init__` sets it when unset, and the property is gone.

One caveat: apply_determinism_to_args still re-derives the fallback by hand (mirror the setting in `__post_init__` ). It runs in validate_args, before the config is built, so there's no resolved value to read yet.

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6967/commits/7e32eefc5ad780a4efc8a7b40771e4119cbe1475

File: megatron/core/transformer/transformer_config.py [hunk 4]
```diff
@@ -971,6 +971,12 @@ class TransformerConfig(ModelParallelConfig):
     supported in TransformerEngine 2.7.0 and above.
     """
 
+    moe_router_aux_loss_fusion: Optional[bool] = None
```
Comment (zhongbozhu) [inline megatron/core/transformer/transformer_config.py:984]: Can we fix the kernel that creates the non-determinism in the first place instead of creating a new toggle that avoids the kernel and guard around it? So if we fix the kernel, we won't need this toggle to exist in the first place?

**Review by zhongbozhu — CHANGES_REQUESTED** [review]

**Review by jaredcasper — APPROVED** [review]

File: megatron/core/transformer/transformer_config.py (hunk 4 above)
Comment (ZhiyuLi-Nvidia) [reply megatron/core/transformer/transformer_config.py:984]: This is a little bit out of the scope of this PR.

I'd propose to merge it for now and later revert the change once we have deterministic loss op in TE and bump up TE to mcore.

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6967/commits/4759068e6d8464ef1e1dd9032f85aacca3d5f455

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6967/commits/ef9b26ca4386b18eec9ca8bb8b2694b520d7d293

File: megatron/core/transformer/transformer_config.py (hunk 4 above)
Comment (ZhiyuLi-Nvidia) [reply megatron/core/transformer/transformer_config.py:984]: Added the issue.

Comment (ZhiyuLi-Nvidia) [reply megatron/core/transformer/transformer_config.py:984]: https://github.com/NVIDIA/Megatron-LM/issues/7221

**Review by zhongbozhu — APPROVED** [review]

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6967/commits/9ba7813b2cceabb67748824a63ab44c7dca3efc1

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6967/commits/1645a1f9d5253441256af2c5b5536872d269c45b

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6967/commits/b6c2fb739133debd24197aa0cff78dc60fee25d7

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6967/commits/ad8d90197cf0d6e6a341fdcb20d167201063ca1f

**Review by yashaswikarnati — APPROVED** [review]
